File: test_lwjgl3/src/Engine/Engine.java
package Engine;

public class Engine implements Runnable{
    private final Thread gameThread;
    private Window window;
    private IGame game;

    public static final int TARGET_FPS = 60;

    public Engine(String windowTitle, int width, int height, boolean vSync, IGame game)
    {
        gameThread = new Thread(this, "GAME_THREAD");
        window = new Window(width, height, windowTitle, vSync);
        this.game = game;
    }

    public void start()
    {
        gameThread.start();
    }

    public void run()
    {
        try{
            init();
            gameLoop();
        } catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    public void init() throws Exception
    {
        window.init();
        game.init();
        System.out.println("--- fin initialisation ---");
    }

    public void gameLoop()
    {
        System.out.println("--- game loop ---");
        boolean running = true;
        while(running && !window.windowShouldClose())
        {
            game.input(window);
            game.update(TARGET_FPS);
            game.render(window);
            window.update();
        }
    }
}

File: openGL/Engine/Engine.cpp
#include "Engine.h"
using namespace std;

Engine* Engine::instance = 0;

Engine::Engine(string windowTitle, int width, int height, bool vSync, Game* g)
{
	window = new Window(width, height, windowTitle, vSync);
	game = g;

	Engine::instance = this;
}

void Engine::run()
{
	init();
	loop();
}

void Engine::init()
{
	window->init();
	game->init();
}

void Engine::loop()
{
	stop = false;

	std::cout << "--- started rendering ---" << std::endl;

	while (stop != true && window->shouldClose() == false)
	{
		game->input(window);
		game->update(TARGET_FPS);
		game->render(window);
		window->update();
	}

	game->flush();

	exit(0);
}

void Engine::shouldStop()
{
	this->stop = true;
	window->setShouldClose(true);
}

Window* Engine::getWindow()
{
	return this->window;
}